csv/src/main/java/com/hazelcast/jet/csv/CsvSource.java

```diff
@@ -0,0 +1,176 @@
+import org.simpleflatmapper.csv.CsvMapperFactory;
+
+import javax.annotation.Nonnull;
+import java.io.BufferedReader;
+import java.io.File;
+import java.io.FileReader;
+import java.io.IOException;
+import java.nio.file.DirectoryStream;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.stream.Stream;
+
+import static com.hazelcast.jet.Traversers.traverseStream;
+import static com.hazelcast.jet.pipeline.Sources.batchFromProcessor;
+
+/**
+ * Reads CSV files and maps lines to Java Beans.
+ *
+ * Please note, that Java Bean should be a public class, in order to work correctly with SimpleFlatMapper.
+ */
+public final class CsvSource<R> extends AbstractProcessor {
+
+    private final Path directory;
+    private final String glob;
+    private final boolean sharedFileSystem;
+
+    private int processorIndex;
+    private int parallelism;
+    private DirectoryStream<Path> directoryStream;
+    private Traverser<R> outputTraverser;
+    private Stream<R> currentStream;
+
+    private final Class<R> itemClass;
+
+    private CsvSource(
+            @Nonnull Class<R> itemClass,
+            @Nonnull String directory,
+            @Nonnull String glob,
+            boolean sharedFileSystem
+    ) {
+        this.itemClass = itemClass;
+        this.directory = Paths.get(directory);
+        this.glob = glob;
+        this.sharedFileSystem = sharedFileSystem;
+    }
+
+    @Override
+    public boolean isCooperative() {
+        return false;
+    }
+
+    @Override
+    protected void init(@Nonnull Context context) throws IOException {
+        processorIndex = sharedFileSystem ? context.globalProcessorIndex() : context.localProcessorIndex();
+        parallelism = sharedFileSystem ? context.totalParallelism() : context.localParallelism();
+
+        directoryStream = Files.newDirectoryStream(directory, glob);
+        outputTraverser = Traversers.traverseIterator(directoryStream.iterator())
+                .filter(this::shouldProcessEvent)
+                .flatMap(this::processFile);
+    }
+
+    @Override
+    public boolean complete() {
+        return emitFromTraverser(outputTraverser);
+    }
+
+    private boolean shouldProcessEvent(Path file) {
+        if (Files.isDirectory(file)) {
+            return false;
+        }
+        int hashCode = file.hashCode();
+        return ((hashCode & Integer.MAX_VALUE) % parallelism) == processorIndex;
+    }
+
+    private Traverser<R> processFile(Path file) {
+        if (getLogger().isFinestEnabled()) {
+            getLogger().finest("Processing file " + file);
+        }
+        assert currentStream == null : "currentStream != null";
+
+        currentStream = createCsvMapper(file);
+        return traverseStream(currentStream)
+                .onFirstNull(() -> {
+                    currentStream.close();
+                    currentStream = null;
+                });
+    }
+
+    @SuppressWarnings("IOResourceOpenedButNotSafelyClosed")
```

Comment:
Instead of suppressing this, I think we can close the file reader in the finally block.

You can use `com.hazelcast.nio.IOUtil#closeResource` utility method to close the resource.

Reply:
does CsvMapper.stream() close the underlying reader when it's closed? Then it's fine since it's closed when it's iterated through the file. We can't close it here.